### PR TITLE
2-D array RGB and RGBA values not understood in plt.plot()

### LINKED ISSUE #18411: 2-D array RGB and RGBA values not understood in plt.plot()
### Bug report

**Bug summary**

2-D array color does not work in plt.plot() while working as intended in plt.scatter(). A 2-D array is recommended by matplotlib itself by this warning:
"'c' argument looks like a single numeric RGB or RGBA sequence, which should be avoided as value-mapping will have precedence in case its length matches with 'x' & 'y'.  Please use a 2-D array with a single row if you really want to specify the same RGB or RGBA value for all points."

**Code for reproduction**

```python
import matplotlib.pyplot as plt
import numpy as np

color_rgb = np.array([[0.1,0.2,0.3]])
color_rgba = np.array([[0.1,0.2,0.3,0.4]])

plt.scatter([1,2],[1,2],c=color_rgb)  # working
plt.plot([1,2],[1,2],c=color_rgb)     # not working

plt.scatter([1,2],[1,2],c=color_rgba) # works fine
plt.plot([1,2],[1,2],c=color_rgba)    # NOT working

plt.show()
```

**Actual outcome**

```
Invalid RGBA argument: array([[0.1, 0.2, 0.3, 0.4]])
```

**Matplotlib version**

  * Operating system: Win 10 1909
  * Matplotlib version: 3.1.3 (from anaconda, pkgs/main)
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.8.1

### PR DESCRIPTION
**Pull request by danuo**
closes #18411

## PR Summary
This fixes the problem discussed in the linked issue in the most simple and general way.

## PR Checklist

- [NA] Has pytest style unit tests (and `pytest` passes).
- [X] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [NA] New features are documented, with examples if plot related.
- [NA] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [X] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [NA] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [NA] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by jklymak** [comment]
A couple of suggestions: 
- say that this "closes #18411" and then it will be automatically closed...
- have a bit more description in the PR. I appreciate you describe the problem in the Issue, but its nicer not to have to click around
- can you add a test?  Otherwise next time this all gets optimized by someone, they will forget this case, and it will need to be fixed again.

**Comment by danuo** [comment]
> 
> 
> A couple of suggestions:
> 
>     * say that this "closes #18411" and then it will be automatically closed...
> 
>     * have a bit more description in the PR. I appreciate you describe the problem in the Issue, but its nicer not to have to click around
> 
>     * can you add a test?  Otherwise next time this all gets optimized by someone, they will forget this case, and it will need to be fixed again.

I am currently trying to create a test. The issue is as follows: The error is produced only when plt.show() is executed. However, it seems that plt.show() never gets executed in any of the existing tests. How would I check for the error when not using plt.show() ?

**Comment by tacaswell** [comment]
`fig.canvas.draw()` will force the rendering.

**Comment by danuo** [comment]
> 
> 
> `fig.canvas.draw()` will force the rendering.

Thank you, it is working great. Currently I am testing for 

3*, 4*,  1D-color-array
1*3, 1*4 2D-color-array

on plot() and scatter() functions. Which other cases do you suppose should be included in the test? I was thinking about hist() for example.

**Comment by tacaswell** [comment]
Using https://matplotlib.org/3.3.0/api/testing_api.html#matplotlib.testing.decorators.check_figures_equal may be very useful here so you can pass the ndim==1 and ndim==2 versions of the same color to the same function and then assert that they generate identical output.

**Comment by jklymak** [comment]
@danuo overall this looks good - be sure to ping us for review if it gets buried...

**Comment by danuo** [comment]
I have all the tests fully working with the `check_figures_equal` decorator. All tests fail with current release and all pass with my commit. Would `test_axes.py` be the right place for these tests? One test looks like this
```Python
@check_figures_equal()
def test_2dcolor_plot(fig_test, fig_ref):
    color = np.array([0.1,0.2,0.3])
    fig_test.subplots().plot([1,2],[1,2],c=color.reshape((-1)))
    fig_ref.subplots().plot([1,2],[1,2], c=color.reshape((1,-1)))
```

**Comment by tacaswell** [comment]
`test_axes` is a reasonable place.  Maybe merge them all into one figure with N axes instead of N individual tests (should run marginally faster)

Probably worth also adding a test in `test_colors` of the conversion function.

I have a slight concern that we are going to find a different case where projecting down here is going to give the wrong answer, but apparently we don't have any test that exercises this code path at all...

**Comment by danuo** [comment]
> Probably worth also adding a test in `test_colors` of the conversion function.

I am not sure what you mean by this. I have implemented all of your other proposals

**Review by jklymak — APPROVED** [review]
Thanks @danuo!

**Comment by tacaswell** [comment]
Putting a test in `test_colors` 

```python
assert mcolors.to_rbga([r, g, b]) == mcolors.to_rgba([[r, g, b]])
```

or similar to directly test the change (rather than implicitly via the plot / scatter behavior).

**Comment by danuo** [comment]
> 
> 
> Putting a test in `test_colors`
> 
> ```python
> assert mcolors.to_rbga([r, g, b]) == mcolors.to_rgba([[r, g, b]])
> ```
> 
> or similar to directly test the change (rather than implicitly via the plot / scatter behavior).

Done, I have implemented such a test in test_colors. Again, the test fails with current matplotlib version and succeeds with this fix.

File: lib/matplotlib/tests/test_axes.py [hunk 1]
```diff
@@ -6513,3 +6513,22 @@ def test_multiplot_autoscale():
...
+    color = np.array([0.1, 0.2, 0.3])
+    # plot with 1D-color:
+    axs = fig_test.subplots(5)
+    axs[0].plot([1, 2], [1, 2], c=color.reshape((-1)))
+    axs[1].scatter([1, 2], [1, 2], c=color.reshape((-1)))
+    axs[2].step([1, 2], [1, 2], c=color.reshape((-1)))
+    axs[3].hist(np.arange(10), color=color.reshape((-1)))
+    axs[4].bar(np.arange(10), np.arange(10), color=color.reshape((-1)))
+    # plot with 2D-color:
+    axs = fig_ref.subplots(5)
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py:6529]: This is backwards, no? The 2D is the new thing to be tested, and the 1D is the reference.

Comment (danuo) [reply lib/matplotlib/tests/test_axes.py:6529]: I agree with your assumption. 1D arrays have a high compatibility due to their similarity to tuples (a,b,c) and lists [a,b,c] while 2D arrays can be misunderstood. Therefore, 1D can be considered the reference.

However, despite the naming of the arguments, isn't this decorator performing a comparison statement (check equal), and therefore this is providing the choice between
`a == b`  and `b == a` ?

Comment (QuLogic) [reply lib/matplotlib/tests/test_axes.py:6529]: Yes, it's just a matter of semantics, but the file names are different, the way it's presented on error is different, etc.

Comment (danuo) [reply lib/matplotlib/tests/test_axes.py]: I have changed the reference accordingly.

File: lib/matplotlib/colors.py [hunk 2]
```diff
@@ -265,6 +265,10 @@ def _to_rgba_no_colorcycle(c, alpha=None):
                     f"Value must be within 0-1 range")
             return c, c, c, alpha if alpha is not None else 1.
         raise ValueError(f"Invalid RGBA argument: {orig_c!r}")
+    # turn 2-D array into 1-D array
+    if isinstance(c, np.ndarray):
+        if c.ndim == 2 and c.shape[0] == 1:
+            c = c.reshape((-1))
```
Comment (QuLogic) [inline lib/matplotlib/colors.py:271]: The inner parentheses do nothing.
```suggestion
            c = c.reshape(-1)
```

File: lib/matplotlib/tests/test_axes.py [hunk 3]
```diff
@@ -6557,3 +6557,22 @@ def test_multiplot_autoscale():
...
+@check_figures_equal(extensions=["pdf"])
+def test_2dcolor_plot(fig_test, fig_ref):
+    color = np.array([0.1, 0.2, 0.3])
+    # plot with 1D-color as reference:
+    axs = fig_ref.subplots(5)
+    axs[0].plot([1, 2], [1, 2], c=color.reshape((-1)))
+    axs[1].scatter([1, 2], [1, 2], c=color.reshape((-1)))
+    axs[2].step([1, 2], [1, 2], c=color.reshape((-1)))
+    axs[3].hist(np.arange(10), color=color.reshape((-1)))
+    axs[4].bar(np.arange(10), np.arange(10), color=color.reshape((-1)))
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_axes.py:6571]: Ditto.
```suggestion
    axs[0].plot([1, 2], [1, 2], c=color.reshape(-1))
    axs[1].scatter([1, 2], [1, 2], c=color.reshape(-1))
    axs[2].step([1, 2], [1, 2], c=color.reshape(-1))
    axs[3].hist(np.arange(10), color=color.reshape(-1))
    axs[4].bar(np.arange(10), np.arange(10), color=color.reshape(-1))
```

File: lib/matplotlib/tests/test_colors.py [hunk 4]
```diff
@@ -1175,3 +1175,10 @@ def test_get_under_over_bad():
     assert_array_equal(cmap.get_under(), cmap(-np.inf))
     assert_array_equal(cmap.get_over(), cmap(np.inf))
     assert_array_equal(cmap.get_bad(), cmap(np.nan))
+
+
+def test_2d_to_rgba():
+    color = np.array([0.1, 0.2, 0.3])
+    rgba_1d = mcolors.to_rgba(color.reshape((-1)))
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_colors.py:1182]:
```suggestion
    rgba_1d = mcolors.to_rgba(color.reshape(-1))
```

**Comment by QuLogic** [comment]
This will need a rebase due to the conflicts.

File: lib/matplotlib/tests/test_colors.py (hunk 4 above)
Comment (danuo) [reply lib/matplotlib/tests/test_colors.py]: I have changed all occurrences of .rebase((-1)) and performed a rebase, please verify.

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
Thanks @danuo! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**